pytorch_finufft/functional.py

```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        None,
+        None,
+        None,
+    ]:
+        """
+        Implements derivatives wrt. each argument in the forward method.
+
+        Parameters
+        ----------
+        ctx : Any
+            Pytorch context object
+        grad_output : torch.Tensor
+            Backpass gradient output.
+
+        Returns
+        -------
+        Tuple[ Union[torch.Tensor, None], ...]
+            A tuple of derivatives wrt. each argument in the forward method
+        """
+        _i_sign = ctx.isign
+        _mode_ordering = ctx.mode_ordering
+        finufftkwargs = ctx.finufftkwargs
+
+        points, targets = ctx.saved_tensors
+        device = points.device
+
+        # start_points = -(np.array(targets.shape) // 2)
+        # end_points = start_points + targets.shape
+        # slices = tuple(slice(start, end) for start, end in zip(start_points, end_points))
+
+        # # CPU idiosyncracy that needs to be done differently
+        # k_ramps = torch.from_numpy(np.mgrid[slices], dtype=points.dtype)
+
+        grad_points = grad_targets = None
+        ndim = points.shape[0]
+
+
+        if ctx.needs_input_grad[0]:
+            # wrt points
+            start_points = -(torch.tensor(targets.shape, device=device) // 2)
+            end_points = start_points + torch.tensor(targets.shape, device=device)
+            coord_ramps = torch.stack(
+                torch.meshgrid(
+                    *(
+                        torch.arange(start, end, device=device)
+                        for start, end in zip(start_points, end_points)
+                    ),
+                    indexing="ij",
+                )
+            )
+
+        if ctx.needs_input_grad[0]:
+            ramped_targets = coord_ramps * targets[np.newaxis] * 1j * _i_sign
+            nufft_func = get_nufft_func(ndim, 2, points.device.type)
+
+            grad_points = nufft_func(
+                    *points,
+                    ramped_targets.squeeze(),
+                    isign=_i_sign,
+                    #modeord=_mode_ordering,
+                    **finufftkwargs,
+                ).conj()  # Currently don't really get why this is hard to replace with a flipped isign
+
+            grad_points = grad_points * grad_output
+            grad_points = torch.atleast_2d(grad_points.real)
+
+        if ctx.needs_input_grad[1]:
+            # wrt. targets
+            nufft_func = get_nufft_func(ndim, 1, points.device.type)
+
+            grad_targets = nufft_func(
+                    *points,
+                    grad_output,
+                    targets.shape,
+                    #modeord=_mode_ordering,
+                    isign=-_i_sign,
+                    **finufftkwargs,
+                )
+            
+            if _mode_ordering == 1:
+                grad_targets = torch.fft.ifftshift(grad_targets, dim=tuple(range(-ndim, 0)))
```

Comment:
Similar note to above on the `dim` argument

Reply:
removed noop stuff